Bug 1325797: Do not fail if the dev-mode pref is missing on startup.

Osmose opened the pull request:
Services.prefs throws when a preference doesn't exist, causing a bug
when, on startup, we check for the dev-mode pref, which happens
before the default value is set for that pref on the first run of the
add-on.

Since Preferences.jsm avoids this issue, it made sense to also port
the rest of the add-on to use it for all preference interactions.

Fixes mozilla/normandy#402.

mythmon commented:
I waffled a bunch on whether to use `Preferences.jsm` or `Services.prefs`. I eventually landed on `Services.prefs` because I wanted the fail-fast behavior in the case we flubbed a key name. I think `Preferences.jsm` will return undefined. Maybe that isn't important though, and we can catch it in code review.

I'm r+ on the change to the dev mode check. I'm unconvinced every where else. What do you think?

Osmose commented:
> I waffled a bunch on whether to use Preferences.jsm or Services.prefs. I eventually landed on Services.prefs because I wanted the fail-fast behavior in the case we flubbed a key name. I think Preferences.jsm will return undefined. Maybe that isn't important though, and we can catch it in code review.

A middle-ground that helps prevent the error but doesn't fail-fast is to centralize preference access for the addon to a single module, and have the other modules import and use that module. I considered playing around with that refactoring but decided not to include it here. What about that?

mythmon commented:
What would that central module look like? `NormandyPrefs.get('dev_mode')`? That doesn't seem to really help us, and it adds layers that make the code harder for a Firefox developer to understand.

I think that centralization that doesn't add fail-fast isn't any better than `Preferences.jsm` in terms of reliability or ease of use, so might as well keep what you have here.

Osmose commented:
> What would that central module look like? NormandyPrefs.get('dev_mode')? That doesn't seem to really help us, and it adds layers that make the code harder for a Firefox developer to understand.

What it gets us is an opportunity to validate against valid preference names and throw if they're incorrect. We don't get the true fail-fast if the preference module itself has incorrect settings names, but it does protect us from typos everywhere else.

I also like that it lets you look in one place to see what preferences are in use by the add-on (and what their defaults are), vs having to hunt down where the defaults are initialized or other actual uses of preferences.

> I think that centralization that doesn't add fail-fast isn't any better than Preferences.jsm in terms of reliability or ease of use, so might as well keep what you have here.

Should I remove the use everywhere else, then? Or leave the PR as it is?

mythmon commented:
Lets go with the minimal change here, just changing the line in the install function. We can figure out the fail loud stuff separately.

mythmon reviewed (changes requested):
Lets go with the minimal change here, just changing the line in the install function. We can figure out the fail loud stuff separately.

Commit:
Bug 1325797: Do not fail if the dev-mode pref is missing on startup.

On first run, the dev-mode pref hasn't been set yet, so the startup
process should have a fallback value to avoid failing in this case.

Osmose commented:
Updated. Since we probably want to merge this soon, I'm going to file an issue for writing a test for the changes to `setDefaultPrefs` separately.
